File: Algoexpert/Find Kth Largest Value In BST/cpp/kth_largest_value.cpp
#include <bits/stdc++.h>

using namespace std;

class BST {
    public:
        int value;
        BST *left = nullptr;
        BST *right = nullptr;

        BST(int value) { this->value = value; }
};

struct TreeInfo {
    int numberOfNodesVisited;
    int latestVistedNodeValue;
};

class Solution {
    public:
        int findKthLargestValueInBstRecursion(BST *tree, int k) {
            vector<int> sortedNodeValues;
            inOrderTraverse(tree, sortedNodeValues);
            return sortedNodeValues[sortedNodeValues.size() - k];
        }

        void inOrderTraverse(BST *node, vector<int> &sortedNodeValues) {
            
            if (node == nullptr) {
                return;
            }

            inOrderTraverse(node->left, sortedNodeValues);
            sortedNodeValues.push_back(node->value);
            inOrderTraverse(node->right, sortedNodeValues);

        }

        int findKthLargestValueInBstReverse(BST *tree, int k) {
            auto treeInfo = TreeInfo{0, -1};
            reverseInOrderTraverse(tree, k, treeInfo);
            return treeInfo.latestVistedNodeValue;
        }

        void reverseInOrderTraverse(BST *node, int k, TreeInfo &treeInfo ) {
            
            if (node == nullptr || treeInfo.numberOfNodesVisited >= k) {
                return;
            }

            reverseInOrderTraverse(node->right, k, treeInfo);
            if(treeInfo.numberOfNodesVisited < k) {
                treeInfo.numberOfNodesVisited++;
                treeInfo.latestVistedNodeValue = node->value;
                reverseInOrderTraverse(node->left, k, treeInfo);
            }
        }
};


int main() {

    auto root = new BST(15);
    root->left = new BST(5);
    root->left->left = new BST(2);
    root->left->left->left = new BST(1);
    root->left->left->right = new BST(3);
    root->left->right = new BST(5);
    root->right = new BST(20);
    root->right->left = new BST(17);
    root->right->right = new BST(22);
    int k = 3;

    Solution solution;
    cout <<  solution.findKthLargestValueInBstRecursion(root, k) << endl;
    cout <<  solution.findKthLargestValueInBstReverse(root, k) << endl;

    return 0;
}

File: Algoexpert/Find Kth Largest Value In BST/python/kth_largest_value.py
import unittest

class BST:
    def __init__(self, value, left=None, right=None):
        self.value = value
        self.left = left
        self.right = right
        
class TreeInfo:
	def __init__(self, numberOfNodesVisited, latestVisitedNodeValue):
		self.numberOfNodesVisited = numberOfNodesVisited
		self.latestVisitedNodeValue = latestVisitedNodeValue

def findKthLargestValueInBstReverse(tree, k):

    treeInfo = TreeInfo(0, -1)
    reverseInOrderTraverse(tree, k, treeInfo)
    return treeInfo.latestVisitedNodeValue

def reverseInOrderTraverse(node, k, treeInfo):
	if node is None or treeInfo.numberOfNodesVisited >= k:
		return
	
	reverseInOrderTraverse(node.right, k, treeInfo)
	if treeInfo.numberOfNodesVisited < k:
		treeInfo.numberOfNodesVisited += 1
		treeInfo.latestVisitedNodeValue = node.value
		reverseInOrderTraverse(node.left, k, treeInfo)

# time O(n) | space O(n)
def findKthLargestValueInBstRecursion(tree, k):
    sortedNodeValues = []
    inOrderTraverse(tree, sortedNodeValues)
    return sortedNodeValues[len(sortedNodeValues) - k]

def inOrderTraverse(node, sortedNodeValues):
    if node is None:
        return
    
    inOrderTraverse(node.left, sortedNodeValues)
    sortedNodeValues.append(node.value)
    inOrderTraverse(node.right, sortedNodeValues)
    
# time O(h) | space O(n)
def findKthLargestValueInBstRecursion(tree, k):
    sortedNodeValues = []
    inOrderTraverse(tree, sortedNodeValues)
    return sortedNodeValues[len(sortedNodeValues) - k]

class BST:
    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None

class TestProgram(unittest.TestCase):
    def test_case_1(self):
        root = BST(15)
        root.left = BST(5)
        root.left.left = BST(2)
        root.left.left.left = BST(1)
        root.left.left.right = BST(3)
        root.left.right = BST(5)
        root.right = BST(20)
        root.right.left = BST(17)
        root.right.right = BST(22)
        k = 3
        expected = 17
        actual = findKthLargestValueInBstRecursion(root, k)
        self.assertEqual(actual, expected)
        
    def test_case_2(self):
        root = BST(15)
        root.left = BST(5)
        root.left.left = BST(2)
        root.left.left.left = BST(1)
        root.left.left.right = BST(3)
        root.left.right = BST(5)
        root.right = BST(20)
        root.right.left = BST(17)
        root.right.right = BST(22)
        k = 3
        expected = 17
        actual = findKthLargestValueInBstReverse(root, k)
        self.assertEqual(actual, expected)

if __name__ == "__main__":
    unittest.main()